pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -575,7 +575,13 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		// need to restart it.
 		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
 			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
-				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
+				var message string
+				if containerStatus == nil {
```

Comment:
looking at `ShouldContainerBeRestarted ` another condition for non dead container to restart may be:

```
// Always restart container in the unknown, or in the created state.
if containerStatus.State == ContainerStateUnknown || containerStatus.State == ContainerStateCreated {
```

Reply:
I think that is out of scope for this PR, I just wanted to fix the log message.

Also, I'm not sure that is correct, IIUC. Can you explain in which case exactly that condition will be met and not handled correctly by the function?

Because the container in unknown state case that you mention, is killed and restarted just a few lines below:
https://github.com/kubernetes/kubernetes/pull/91469/files#diff-8da51ab012ecbdf2b013ecbcf1c37033R587

And if the container is not running, is recycled a few lines above: https://github.com/kubernetes/kubernetes/pull/91469/files#diff-8da51ab012ecbdf2b013ecbcf1c37033R567. And, IIUC, will be restarted (if needs to be restarted) in following calls to this function.

Am I missing something?

Also, if you think it is okay, can you open an issue to discuss that case and keep it out of scope for this simple PR?

Reply:
I'm new to codebase and commenting with little context. So when the status was `containerStatus.State == ContainerStateCreated`, do you also want to see error message about container that was dead. Or you are saying it will be recycled by `m.internalLifecycle.PostStopContainer` and `ShouldContainerBeRestarted` will not return `true` to enter this `if` statement? Or it is totally impossible to have `containerStatus.State == ContainerStateCreated` in the beginning of this loop cycle at a first place? Thank you for your reply!

Reply:
@SergeyKanzhelev oh, I think I now understand your first message. Sorry.

Yes, maybe in those cases the dead container is not the best message either. Something like @derekwaynecarr mentioned seems better:
    `Container %q of pod %q is not in desired state and shall be started`

What do you think?

Reply:
You can also include the `containerStatus` in the message and there will be no need to increase verbosity of `ShouldContainerBeRestarted`.

Reply:
I'll update as derek mentioned here: https://github.com/kubernetes/kubernetes/pull/91469#discussion_r465145008

Marking this as solved, but don't hesitate to let me know if I'm missing something! :)